daemon/daemon.go

```diff
@@ -1515,6 +1515,34 @@ func CreateDaemonRoot(config *config.Config) error {
 	return setupDaemonRoot(config, realRoot, idMapping.RootPair())
 }
 
+// RemapContainerdNamespaces returns the right containerd namespaces to use:
+// - if they are not already set in the config file
+// -  and the daemon is running with user namespace remapping enabled
+// Then it will return new namespace names, otherwise it will return the existing
+// namespaces
+func RemapContainerdNamespaces(config *config.Config) (ns string, pluginNs string, err error) {
+	idMapping, err := setupRemappedRoot(config)
+	if err != nil {
+		return "", "", err
+	}
+	if idMapping.Empty() {
+		return config.ContainerdNamespace, config.ContainerdPluginNamespace, nil
+	}
+	root := idMapping.RootPair()
+
+	ns = config.ContainerdNamespace
+	if _, ok := config.ValuesSet["containerd-namespace"]; !ok {
+		ns = fmt.Sprintf("%s-%d.%d", config.ContainerdNamespace, root.UID, root.GID)
+	}
+
+	pluginNs = config.ContainerdPluginNamespace
+	if _, ok := config.ValuesSet["containerd-plugin-namespace"]; !ok {
+		pluginNs = fmt.Sprintf("%s-%d.%d", config.ContainerdPluginNamespace, root.UID, root.GID)
+	}
```

Comment:
There might be a usability issue;  if someone specifies the namespaces in their config, we will use those. Which means that if someone specified a custom namespace, and that namespace is _not_ empty, things will likely start failing;

```
     --containerd-namespace string             Containerd namespace to use (default "moby")
      --containerd-plugins-namespace string     Containerd namespace to use for plugins (default "plugins.moby")
```

Wondering if there's a way for us to detect;

- if the namespace is non-empty
- if it's not empty, if it was created with the correct user-mapping

And to otherwise bail out and produce an error

(Should be fine for a follow-up, but maybe we should create a tracking ticket?)

Reply:
My idea was that, if they tell us which namespace to use then they know what they are doing. Now yeah, the namespace could have an incorrect mapping, we should do something about that, we could add a label to the namespace (looks like that's possible in containerd), but then that would work only for the ones we create, I'm not sure how we can check that an existing namespace has the right user-mapping... Maybe we can make it work only with the namespaces _we_ create? Ok let's open an issue for this, a comment in a PR is not the best place to find a solution for this

Reply:
Yes, it's not a blocker for this PR, but something to look at. My consideration here was;

1. user configures daemon (with namespace)
2. runs the daemon with that config
3. user now decides to enable user-namespaces
4. daemon now fails with potentially confusing errors
5. and.. now how to solve? (how to clean the namespace etc?)